src/modules/peek/Peek.UI/FileQuery.cs

```diff
@@ -0,0 +1,170 @@
+    using Peek.Common.Models;
+    using Peek.UI.Helpers;
+
+    public partial class FileQuery : ObservableObject
+    {
+        private const int UninitializedItemIndex = -1;
+
+        public void Uninitialize()
+        {
+            CurrentFile = null;
+
+            if (initializeFilesTask != null && initializeFilesTask.Status == TaskStatus.Running)
+            {
+                Debug.WriteLine("Detected existing initializeFilesTask running. Cancelling it..");
+                cancellationTokenSource.Cancel();
+            }
+
+            initializeFilesTask = null;
+
+            lock (mutateQueryDataLock)
+            {
+                files = new List<File>();
+                currentItemIndex = UninitializedItemIndex;
+            }
+        }
+
+        public void UpdateCurrentItemIndex(int desiredIndex)
+        {
+            if (files == null || files.Count <= 1 || currentItemIndex == UninitializedItemIndex ||
+                (initializeFilesTask != null && initializeFilesTask.Status == TaskStatus.Running))
+            {
+                return;
+            }
+
+            // Current index wraps around when reaching min/max folder item indices
+            desiredIndex %= files.Count;
+            currentItemIndex = desiredIndex < 0 ? files.Count + desiredIndex : desiredIndex;
+
+            if (currentItemIndex < 0 || currentItemIndex >= files.Count)
+            {
+                Debug.Assert(false, "Out of bounds folder item index detected.");
+                currentItemIndex = 0;
+            }
+
+            CurrentFile = files[currentItemIndex];
+        }
+
+        public void Initialize()
+        {
+            var folderView = FileExplorerHelper.GetCurrentFolderView();
+            if (folderView == null)
+            {
+                return;
+            }
+
+            Shell32.FolderItems selectedItems = folderView.SelectedItems();
+            if (selectedItems == null || selectedItems.Count == 0)
+            {
+                return;
+            }
+
+            // Prioritize setting CurrentFile, which notifies UI
+            var firstSelectedItem = selectedItems.Item(0);
+            CurrentFile = new File(firstSelectedItem.Path);
+
+            var items = selectedItems.Count > 1 ? selectedItems : folderView.Folder?.Items();
+            if (items == null)
+            {
+                return;
+            }
+
+            try
+            {
+                if (initializeFilesTask != null && initializeFilesTask.Status == TaskStatus.Running)
+                {
+                    Debug.WriteLine("Detected unexpected existing initializeFilesTask running. Cancelling it..");
+                    cancellationTokenSource.Cancel();
+                }
+
+                cancellationTokenSource = new CancellationTokenSource();
+                initializeFilesTask = new Task(() => InitializeFiles(items, firstSelectedItem, cancellationTokenSource.Token));
```

Comment:
@miksalmon should we just use Task.Run and not myTask.Start(). I rarely see Start() being used.

Reply:
Seems to be safe, based on what I've seen in the docs - just fires and forgets on a worker thread, unless we await it. We can always swap it out later if it becomes problematic for whatever reason.